op-service/sync/scheduler.go

```diff
@@ -0,0 +1,77 @@
+package sync
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+// ErrChannelFull is returned when the scheduler's processing channel is full
+// and a new item cannot be scheduled.
+var ErrChannelFull = errors.New("channel full")
+
+// SchedulerRunner is a function that processes items received by the [Scheduler].
+// Errors returned by the runner are sent to the [SchedulerErrors] channel.
+// The user of the [Scheduler] is responsible for handling these errors using the [SchedulerErrors] receiving channel.
+type SchedulerRunner[T any] func(ctx context.Context, item T) error
+
+// SchedulerRunnerErrorHandler is a callback function
+// that is invoked when the [SchedulerRunner] errors.
+type SchedulerRunnerCallback[T any] func(item T, err error)
+
+// Scheduler processes generic [T any] items using a provided runner function.
+// Processing may be buffered with the [NewSchedulerFromBufferSize] constructor.
+type Scheduler[T any] struct {
+	receiver chan T
+	runner   SchedulerRunner[T]
+	callback SchedulerRunnerCallback[T]
+	cancel   func()
+	wg       sync.WaitGroup
+}
+
+func NewSchedulerFromBufferSize[T any](runner SchedulerRunner[T], callback SchedulerRunnerCallback[T], bufferSize int) *Scheduler[T] {
+	return &Scheduler[T]{
+		receiver: make(chan T, bufferSize),
+		runner:   runner,
+		callback: callback,
+	}
+}
+
+// Start starts the scheduler.
+func (s *Scheduler[T]) Start(ctx context.Context) {
+	ctx, cancel := context.WithCancel(ctx)
+	s.cancel = cancel
+	s.wg.Add(1)
+	go s.run(ctx)
+}
+
+// Close stops the scheduler and waits for all in-flight processing to finish.
+func (s *Scheduler[T]) Close() error {
+	s.cancel()
+	s.wg.Wait()
+	return nil
+}
+
+func (s *Scheduler[T]) run(ctx context.Context) {
+	defer s.wg.Done()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case item := <-s.receiver:
+			if err := s.runner(ctx, item); err != nil {
+				s.callback(item, err)
```

Comment:
Would it make more sense to just not have the runner return an error?  We could still use the existing functions like verifying a preimage that returns an error, but instead of passing the error handling as a separate function, it would just be a wrapper around the runner function.  ie with the callback we'd have pseudo code like:

```
NewScheduler(runner, func(err) { 
  log.Error(err) 
})
```

But we could just have:

```
NewScheduler(func() {
  if err := runner(); err != nil {
    log.Error(err)
  }
})
```

Reply:
This totally makes sense. I think I initially thought I would need something local to each scheduler's implementation for the callback, but it seems there's nothing additional needed besides a logger 😄

Changed it up to follow your suggested pattern!